### PR TITLE
Inbound Transaction queue processing

### LINKED ISSUE #671: S7 [User Story] Add transactions payloads to inbound queue
[User Story] | Add import transaction payloads to inbound queue
----------|----------
Epic Title | #333 
User Story Title | Add user transactions payloads to inbound queue
T-shirt size: S(mall), M(edium), L(arge) | M
Story Points - Fibonacci (1,2,3,5,8,13) | 5
User Story | Add transactions payloads to inbound queue

- [ ] Add User transactions 
- [ ] Add group transactions

### PR DESCRIPTION
**Pull request by adamgering**
Adds transaction data to inbound queue records
Reads inbound queue and sends to validator

Resolves #671 and #673 for inbound initial sync, LDAP and Azure.

### CODE REVIEW

File: rbac/ledger_sync/inbound/listener.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                break
+            LOGGER.info("Processed %s records in the inbound queue", count)
+        LOGGER.info("Listening for incoming Sawtooth transactions")
+        feed = database.run_query(database.get_table("inbound_queue").changes())
+        for rec in feed:
+            if rec["new_val"] and not rec["old_val"]:  # only insertions
+                process(rec["new_val"], database)
+
+    except Exception as err:  # pylint: disable=broad-except
+        LOGGER.exception("Outbound listener %s exception", type(err).__name__)
```
Comment (kthblmfld) [inline rbac/ledger_sync/inbound/listener.py]: This should be Inbound listener %s exception

File: rbac/providers/ldap/initial_inbound_sync.py [hunk 2]
```diff
@@ -93,7 +95,9 @@ def insert_to_db(data_dict, data_type):
             "sync_type": "initial",
             "timestamp": datetime.now().replace(tzinfo=timezone.utc).isoformat(),
             "provider_id": LDAP_DC,
+            "raw": entry_data,
         }
+        add_transaction(inbound_entry)
```
Comment (nadiabahrami) [inline rbac/providers/ldap/initial_inbound_sync.py]: This is adding transaction information prior to being loaded into the inbound queue.  I was under the impression that this sort of processing would be done post insertion into the inbound queue.  This seems to be in the wrong part of the flow as I thought this is supposed to go into the NEXT portion of the application.

Comment (kthblmfld) [reply rbac/providers/ldap/initial_inbound_sync.py]: I am also confused about this. The queues are the handoff points between modules. inbound_sync should be publishing to this queue and ledger_sync should be subscribing.

Comment (kthblmfld) [reply rbac/providers/ldap/initial_inbound_sync.py]: Please relocate this operation to the inbound processing step of ledger_sync. 
Ldap inbound enqueues the records, ledger_sync dequeues and processes them.

**Comment by kthblmfld** [comment]
I just noticed this doesn't include outbound_queue publication for AD/AAD, but gathered from yesterday's discussions that it is there. If it isn't in NEXT, we will need to open a story to build it to complete the data flow through the system.

File: rbac/providers/common/provider_transforms.py [hunk 3]
```diff
@@ -12,9 +12,11 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 # ------------------------------------------------------------------------------
+""" Transforms incoming data based on key names and provider type
+"""
 
 USER_TRANSFORM = {
-    "user_id": {"azure": "id", "ldap": "objectGUID"},
+    "user_id": {"azure": "id", "ldap": "distinguishedName"},
```
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/providers/common/provider_transforms.py]: Can you explain why objectGUID is correct?

File: rbac/providers/common/inbound_filters.py [hunk 4]
```diff
@@ -12,23 +12,37 @@
...
-        if value[provider] in user:
-            clean_user[key] = user[value[provider]]
-        else:
-            clean_user[key] = None
+    for key, alias in USER_TRANSFORM.items():
+        if alias[provider] in user:
+            value = inbound_value_filter(user[alias[provider]])
+            if value:
+                clean_user[key] = value
+    for key, aliases in STANDARD_USER_TRANSFORM.items():
```
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/providers/common/inbound_filters.py]: Name is required by NEXT, but it is NULL on many records in LDAP. Thus we fall back to those other fields if NAME is NULL.

I will query for a list of records that have no name field in LDAP.

File: rbac/providers/common/provider_transforms.py [hunk 5]
```diff
@@ -48,7 +50,7 @@
 }
 
 GROUP_TRANSFORM = {
-    "role_id": {"azure": "id", "ldap": "objectGUID"},
+    "role_id": {"azure": "id", "ldap": "distinguishedName"},
```
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/providers/common/provider_transforms.py]: Can you explain why objectGUID is the appropriate UUID field for a group, versus distinguishedName?

File: rbac/ledger_sync/inbound/listener.py (hunk 1 above)
Comment (adamgering) [reply rbac/ledger_sync/inbound/listener.py]: resolved

File: rbac/providers/common/provider_transforms.py [hunk 6]
```diff
@@ -97,3 +98,7 @@
     "owners": {"azure": "owners", "ldap": "managedBy"},
     "visibility": {"azure": "visibility", "ldap": None},
 }
+
+STANDARD_USER_TRANSFORM = {"name": {"name", "given_name", "user_nickname"}}
+
+STANDARD_GROUP_TRANSFORM = {}
```
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/providers/common/provider_transforms.py]: resolved

File: rbac/providers/ldap/initial_inbound_sync.py (hunk 2 above)
Comment (adamgering) [reply rbac/providers/ldap/initial_inbound_sync.py]: resolved

File: tests/unit/providers/common/outbound_filter_tests.py [hunk 7]
```diff
@@ -76,7 +76,7 @@ def test_outbound_user_filter_bad_provider():
 def test_outbound_group_filter():
     """ Test outbound group filter with valid user """
     result = outbound_group_filter({"role_id": 1234}, "ldap")
-    assert result["objectGUID"] == 1234
+    assert result["distinguishedName"] == 1234
```
[the remark this answers is not in the document]
Comment (adamgering) [reply tests/unit/providers/common/outbound_filter_tests.py]: See prior comments

File: rbac/providers/common/inbound_filters.py [hunk 8]
```diff
@@ -37,12 +51,33 @@ def inbound_group_filter(group, provider):
...
-        if value[provider] in group:
-            clean_group[key] = group[value[provider]]
-        else:
-            clean_group[key] = None
+    for key, alias in GROUP_TRANSFORM.items():
+        if alias[provider] in group:
+            value = inbound_value_filter(group[alias[provider]])
+            if value:
+                clean_group[key] = value
+    for key, aliases in STANDARD_GROUP_TRANSFORM.items():
```
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/providers/common/inbound_filters.py]: resolved

**Comment by adamgering** [comment]
> I just noticed this doesn't include outbound_queue publication for AD/AAD, but gathered from yesterday's discussions that it is there, and from last Sprint's commitments and our agreement Friday that it should be done and included here (otherwise I would have finished it out over the weekend and not deleted the work @jbobo and I had under the pre-Friday architecture). If it isn't in NEXT, we will need to open a story to build it to complete the data flow through the system.

Do you have a Issue number for outbound sync?

File: rbac/providers/common/inbound_filters.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/providers/common/inbound_filters.py]: There are 67 records out of 907 records in DC=clouddev that have no value for [name].

File: rbac/providers/ldap/initial_inbound_sync.py [hunk 9]
```diff
@@ -93,7 +96,10 @@ def insert_to_db(data_dict, data_type):
             "sync_type": "initial",
             "timestamp": datetime.now().replace(tzinfo=timezone.utc).isoformat(),
             "provider_id": LDAP_DC,
+            "raw": entry_data,
         }
+        # TODO: add_transaction will likely need to be included here for security purposes
```
Comment (kthblmfld) [inline rbac/providers/ldap/initial_inbound_sync.py]: Uh oh. We should discuss the implications w/respect to the providers.

File: rbac/providers/common/provider_transforms.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/providers/common/provider_transforms.py]: Reverted

File: rbac/providers/common/provider_transforms.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/providers/common/provider_transforms.py]: Reverted

File: tests/unit/providers/common/outbound_filter_tests.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (adamgering) [reply tests/unit/providers/common/outbound_filter_tests.py]: Reverted

File: rbac/providers/common/inbound_filters.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamgering) [reply rbac/providers/common/inbound_filters.py]: Leaving per info above.

**Comment by kthblmfld** [comment]
FollowUp: @adamgering, Dustin, and I reviewed the architecture here and have agreed to move the ObjectGuid to a NEXT entity field called 'remote_id'. Having the distinguishedName mapped to user_id maintains the referential integrity of state records.

This should allow consistent import/export of data to/from LDAP as well as consistent import/export of data to/from Sawtooth. @nadiabahrami @mtn217 please chime in quickly if there are any outstanding issues with this approach. We can always revisit and adjust if and when side effects occur.

**Comment by mtn217** [comment]
Will the 'id' field from Azure AD be mapped to the NEXT identity field 'remote_id'?

Michael
________________________________

**Comment by adamgering** [comment]
superceded by PR #818

**Pull request closed without merging**